### ISSUE TITLE
NavMap::sync `write_lock` blocks main loop when no write is necessary, causes unnecessary stuttering

### ISSUE THREAD

**Issue by cgodley** [godotengine/godot#96820]
### Tested versions

- Reproducible in v4.3.stable
- `write_lock` was first introduced in commit `4cc8748`
- Issue is resolved if I compile engine with `write_lock` removed

### System information

Godot v4.3.stable - macOS 14.6.1 - Vulkan (Forward+) - integrated Apple M1 Pro - Apple M1 Pro (8 Threads)

### Issue description

My game frequently calls `NavigationServer3D.map_get_path()`. I use separate threads (`WorkerThreadPool.add_task()`) to avoid stuttering caused by blocking the main loop.

Beginning with git hash `4cc8748`, my game frequently has significant stutters. This is because `NavMap::sync()` (in the main loop) acquires `write_lock` every single iteration, and so the main loop cannot proceed until _all threads_ are done with `read_lock`. This seems unnecessarily slow. Instead, perhaps we could detect when a write will be required, and only then acquire the `write_lock`. Or perhaps someone has a better suggestion.

I've tried using simpler collision shapes, tagging specific areas as "requiring navigation" vs not requiring it, performing less frequent navigation, etc. but it has not been sufficient to fully resolve the stuttering. I suspect this issue has at least some effect on the frame time of _all games_ using threaded navigation.

I think I have confirmed the cause by testing the MRP (attached) before/after recompiling Godot with `write_lock` removed. Before the change there is obvious stuttering. After removing `write_lock`, everything runs smoothly.

### Steps to reproduce

Run the MRP, you will see a smoothly animating sphere for a few seconds. Then it will schedule some expensive navigation tasks in a low priority `WorkerThreadPool`. On affected versions of Godot, you will observe the framerate drop significantly while navigation is running despite the use of `WorkerThreadPool`.

### Minimal reproduction project (MRP)

[navmap_stutter.zip](https://github.com/user-attachments/files/16949854/navmap_stutter.zip)

**Comment by cgodley** [comment]
Below is the diff that works in my local testing to resolve the issue (without removing the lock).

I'd appreciate it if someone could review the changes since this is my first time contributing to Godot.

```diff
diff --git a/modules/navigation/nav_map.cpp b/modules/navigation/nav_map.cpp
index f917c988ea..bef18d9f77 100644
--- a/modules/navigation/nav_map.cpp
+++ b/modules/navigation/nav_map.cpp
@@ -896,7 +896,29 @@ Vector3 NavMap::get_random_point(uint32_t p_navigation_layers, bool p_uniformly)
 }
 
 void NavMap::sync() {
-	RWLockWrite write_lock(map_rwlock);
+
+    // Check if lock is necessary
+    bool region_dirty = false;
+	for (NavRegion *region : regions) {
+		if (region->is_dirty()) {
+			region_dirty = true;
+		}
+	}
+	for (NavLink *link : links) {
+		if (link->check_dirty()) {
+            regenerate_links = true;
+		}
+	}
+	for (NavObstacle *obstacle : obstacles) {
+		if (obstacle->check_dirty()) {
+			obstacles_dirty = true;
+		}
+	}
+	for (NavAgent *agent : agents) {
+		if (agent->check_dirty()) {
+			agents_dirty = true;
+		}
+	}
 
     // Performance Monitor
     int _new_pm_region_count = regions.size();
@@ -908,6 +930,11 @@ void NavMap::sync() {
     int _new_pm_edge_connection_count = pm_edge_connection_count;
     int _new_pm_edge_free_count = pm_edge_free_count;
 
+    bool need_write = regenerate_polygons || region_dirty || regenerate_links || obstacles_dirty || agents_dirty;
+    if (need_write) {
+
+        RWLockWrite write_lock(map_rwlock);
+
         // Check if we need to update the links.
         if (regenerate_polygons) {
             for (NavRegion *region : regions) {
@@ -1204,6 +1231,7 @@ void NavMap::sync() {
         if (obstacles_dirty || agents_dirty) {
             _update_rvo_simulation();
         }
+    }
 
 	regenerate_polygons = false;
 	regenerate_links = false;
diff --git a/modules/navigation/nav_region.cpp b/modules/navigation/nav_region.cpp
index f30855d697..d27110667a 100644
--- a/modules/navigation/nav_region.cpp
+++ b/modules/navigation/nav_region.cpp
@@ -206,6 +206,10 @@ Vector3 NavRegion::get_random_point(uint32_t p_navigation_layers, bool p_uniform
 	}
 }
 
+bool NavRegion::is_dirty() {
+    return polygons_dirty;
+}
+
 bool NavRegion::sync() {
 	bool something_changed = polygons_dirty /* || something_dirty? */;
 
diff --git a/modules/navigation/nav_region.h b/modules/navigation/nav_region.h
index ebc082bd2f..6098db1c9a 100644
--- a/modules/navigation/nav_region.h
+++ b/modules/navigation/nav_region.h
@@ -100,6 +100,7 @@ public:
 
 	real_t get_surface_area() const { return surface_area; };
 
+    bool is_dirty();
 	bool sync();
 
 private:
```

**Comment by cgodley** [comment]
Will open a pull request soon to demonstrate one way to fix it

**Comment by firescaleVR** [comment]
This is terrible. Polling any unreachable place in the navmesh causes it to stall for 300-500milliseconds, and now thanks to 4cc8748 the main loop freezes making any future versions of godot unuseable for me.
